controllers/awscluster_controller_unit_test.go

```diff
@@ -0,0 +1,637 @@
+			awsCluster: &infrav1.AWSCluster{
+				ObjectMeta: metav1.ObjectMeta{Name: "aws-test-1", Namespace: "default"}, TypeMeta: metav1.TypeMeta{Kind: "AWSCluster", APIVersion: infrav1.GroupVersion.String()},
+			},
+			ownerCluster: &clusterv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{Name: "capi-test", Namespace: "default"}, Spec: clusterv1.ClusterSpec{InfrastructureRef: &corev1.ObjectReference{
+					APIVersion: infrav1.GroupVersion.String(),
+					Kind:       "AWSCluster",
+					Name:       "aws-test-1",
+					Namespace:  "default",
+				}}},
+			requeue: true,
+		},
+		{
+			name: "Should not create reconcile request if AWSCluster is externally managed",
+			awsCluster: &infrav1.AWSCluster{
+				ObjectMeta: metav1.ObjectMeta{Name: "aws-test-2", Namespace: "default", Annotations: map[string]string{clusterv1.ManagedByAnnotation: "capi-test"}},
+				TypeMeta:   metav1.TypeMeta{Kind: "AWSCluster", APIVersion: infrav1.GroupVersion.String()},
+			},
+			ownerCluster: &clusterv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{Name: "capi-test", Namespace: "default"}, Spec: clusterv1.ClusterSpec{InfrastructureRef: &corev1.ObjectReference{
+					APIVersion: infrav1.GroupVersion.String(),
+					Kind:       "AWSCluster",
+					Name:       "aws-test-2",
+					Namespace:  "default",
+				}}},
+			requeue: false,
+		},
+		{
+			name:         "Should not create reconcile request for deleted clusters",
+			ownerCluster: &clusterv1.Cluster{ObjectMeta: metav1.ObjectMeta{Name: "capi-test", Namespace: "default", DeletionTimestamp: &metav1.Time{Time: time.Now()}}},
+			requeue:      false,
+		},
+		{
+			name:         "Should not create reconcile request if infrastructure ref for AWSCluster on owner cluster is not set",
+			ownerCluster: &clusterv1.Cluster{ObjectMeta: metav1.ObjectMeta{Name: "capi-test", Namespace: "default"}},
+			requeue:      false,
+		},
+		{
+			name: "Should not create reconcile request if infrastructure ref type on owner cluster is not AWSCluster",
+			ownerCluster: &clusterv1.Cluster{ObjectMeta: metav1.ObjectMeta{Name: "capi-test", Namespace: "default"}, Spec: clusterv1.ClusterSpec{InfrastructureRef: &corev1.ObjectReference{
+				APIVersion: clusterv1.GroupVersion.String(),
+				Kind:       "Cluster",
+				Name:       "aws-test"}}},
+			requeue: false,
+		},
+		{
+			name: "Should not create reconcile request if AWSCluster not found",
+			ownerCluster: &clusterv1.Cluster{ObjectMeta: metav1.ObjectMeta{Name: "capi-test", Namespace: "default"}, Spec: clusterv1.ClusterSpec{InfrastructureRef: &corev1.ObjectReference{
+				APIVersion: clusterv1.GroupVersion.String(),
+				Kind:       "AWSCluster",
+				Name:       "aws-test"}}},
+			requeue: false,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			g := NewWithT(t)
+			ctx := context.Background()
+			log := ctrl.LoggerFrom(ctx)
+			reconciler := &AWSClusterReconciler{
+				Client: testEnv.Client,
+			}
+			if tc.awsCluster != nil {
```

Comment:
actually above lines contain defer statement which should be executed after the func execution completes.
If i extract it to other func, it will change the defer behaviour. That's why not extracted to other func

Reply:
What if  we do something similar to this

```
func cleanupCluster(g *WithT, awsCluster *infrav1.AWSCluster) {
	if awsCluster !=nil{
		g.Expect(testEnv.Cleanup(ctx, awsCluster)).To(Succeed())
	}
}
func setupCluster(g *WithT, awsCluster *infrav1.AWSCluster) {
	if awsCluster != nil {
		awsCluster.Default()
		g.Expect(testEnv.Create(ctx, awsCluster)).To(Succeed())
	}
}
```

And in your test cases just call 
```
setupCluster(g, tc.awsCluster)
defer cleanupCluster(g, tc.awsCluster)
```

Reply:
I can do that 👍